pkg/controllermanager/controller/controllerregistration/extensionclusterrole/reconciler.go

```diff
@@ -0,0 +1,126 @@
+	clusterRole.SetGroupVersionKind(rbacv1.SchemeGroupVersion.WithKind("ClusterRole"))
+	if err := r.Client.Get(ctx, request.NamespacedName, clusterRole); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.V(1).Info("Object is gone, stop reconciling")
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, fmt.Errorf("error retrieving object from store: %w", err)
+	}
+
+	if clusterRole.DeletionTimestamp != nil {
+		return reconcile.Result{}, nil
+	}
+
+	subjects, err := r.computeSubjects(ctx, clusterRole)
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("failed computing subjects for ClusterRoleBinding: %w", err)
+	}
+
+	clusterRoleBinding := &rbacv1.ClusterRoleBinding{ObjectMeta: metav1.ObjectMeta{Name: clusterRole.Name}}
+	_, err = controllerutils.GetAndCreateOrMergePatch(ctx, r.Client, clusterRoleBinding, func() error {
+		ownerReference := metav1.NewControllerRef(clusterRole, rbacv1.SchemeGroupVersion.WithKind("ClusterRole"))
+		ownerReference.BlockOwnerDeletion = pointer.Bool(false)
+		clusterRoleBinding.SetOwnerReferences(kubernetesutils.MergeOwnerReferences(clusterRoleBinding.GetOwnerReferences(), *ownerReference))
+
+		clusterRoleBinding.RoleRef = rbacv1.RoleRef{
+			APIGroup: rbacv1.SchemeGroupVersion.Group,
+			Kind:     "ClusterRole",
+			Name:     clusterRole.Name,
+		}
+		clusterRoleBinding.Subjects = subjects
+		return nil
+	})
+	return reconcile.Result{}, err
+}
+
+func (r *Reconciler) computeSubjects(ctx context.Context, clusterRole *metav1.PartialObjectMetadata) ([]rbacv1.Subject, error) {
+	extensionNames := strings.Split(clusterRole.Annotations[v1beta1constants.LabelExtensionsAuthorizationBindTo], ",")
+	if len(extensionNames) == 0 {
+		return nil, nil
+	}
+
+	seedNamespaceList := &metav1.PartialObjectMetadataList{}
+	seedNamespaceList.SetGroupVersionKind(corev1.SchemeGroupVersion.WithKind("NamespaceList"))
+	if err := r.Client.List(ctx, seedNamespaceList, client.MatchingLabels{v1beta1constants.GardenRole: v1beta1constants.GardenRoleSeed}); err != nil {
+		return nil, fmt.Errorf("failed listing seed namespaces: %w", err)
+	}
+
+	// ensure stable order of subjects
+	kubernetesutils.ByName().Sort(seedNamespaceList)
+
+	var subjects []rbacv1.Subject
+
+	for _, namespace := range seedNamespaceList.Items {
+		serviceAccountList := &metav1.PartialObjectMetadataList{}
+		serviceAccountList.SetGroupVersionKind(corev1.SchemeGroupVersion.WithKind("ServiceAccountList"))
+		if err := r.Client.List(ctx, serviceAccountList, client.InNamespace(namespace.Name)); err != nil {
+			return nil, fmt.Errorf("failed listing ServiceAccounts in namespace %s: %w", namespace.Name, err)
+		}
+
+		// ensure stable order of subjects
+		kubernetesutils.ByName().Sort(serviceAccountList)
+
+		for _, serviceAccount := range serviceAccountList.Items {
+			for _, extensionName := range extensionNames {
+				if strings.HasPrefix(serviceAccount.Name, v1beta1constants.ExtensionGardenServiceAccountPrefix+extensionName) {
```

Comment:
I noticed that matching the `extensionName` has some drawbacks:
- An extension cannot create multiple `ServiceAccount`s with distinct permissions, since the permissions are aggregated.
- We calculate `ClusterRoleBinding`s again and again due to the `ServiceAccount` watch, i.e. it has no direct relation to the desired `ClusterRole`.

How about if we don't use the extension but the service account name for matching?

```
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: extension-create-subjectaccessreviews
  annotations:
    authorization.extensions.gardener.cloud/bind-to: <service-account-name>
```

Reply:
The problem with using the name of the service account is that it is not static/known upfront. `gardenlet` uses the extension namespace name in its seed for the service account name, and this is different in each seed/for each installation.

TBH, I don't have good answers for the points above (especially the first one). I guess this is somewhat related to the concept of reducing permissions (which we don't have today).

Reply:
Sure, but this also means that any reasonable usage of this feature requires to create new `ServiceAccount`s just because the name needs to be static. I'm not sure whether that's desireable.

Reply:
> Sure, but this also means that any reasonable usage of this feature requires to create new ServiceAccounts just because the name needs to be static. I'm not sure whether that's desireable.

That was exactly the reason for #8980: Extensions request a separate and dedicated service account for a component that is additionally deployed by them to the seed. It was actually never intended to re-use the existing one.